lib/pubnub/events/grant_token.rb

```diff
@@ -0,0 +1,195 @@
+        event: @event,
+        event_options: @given_options,
+        timetoken: nil,
+        status: {
+          code: req_res_objects[:response].code,
+          client_request: req_res_objects[:request],
+          server_response: req_res_objects[:response],
+          category: Pubnub::Constants::STATUS_ACK,
+          error: false,
+          auto_retried: false,
+
+          current_timetoken: nil,
+          last_timetoken: nil,
+          subscribed_channels: nil,
+          subscribed_channel_groups: nil,
+
+          data: nil,
+
+          config: get_config
+
+        },
+        result: {
+          code: req_res_objects[:response].code,
+          operation: current_operation,
+          client_request: req_res_objects[:request],
+          server_response: req_res_objects[:response],
+
+          data: parsed_response['data']
+        }
+      )
+    end
+
+    def path
+      '/' + [
+        'v3',
+        'pam',
+        @subscribe_key,
+        'grant'
+      ].join('/')
+    end
+  end
+
+  class Permissions
+    attr_reader :type
+
+    class << Permissions
+      def pat(read: false, write: false, manage: false, delete: false,
+              create: false, get: false, update: false, join: false)
+        Permissions.new(
+          type: :pattern,
+          read: read,
+          write: write,
+          manage: manage,
+          delete: delete,
+          create: create,
+          get: get,
+          update: update,
+          join: join
+        )
+      end
+
+      def res(read: false, write: false, manage: false, delete: false, create: false, get: false, update: false, join: false)
+        Permissions.new(
+          type: :resource,
+          read: read,
+          write: write,
+          manage: manage,
+          delete: delete,
+          create: create,
+          get: get,
+          update: update,
+          join: join
+        )
+      end
+    end
+
+    def initialize(options)
+      @type = options[:type]
+      @read = options[:read] || false
+      @write = options[:write] || false
+      @manage = options[:manage] || false
+      @delete = options[:delete] || false
+      @create = options[:create] || false
+      @get = options[:get] || false
+      @update = options[:update] || false
+      @join = options[:join] || false
+    end
+
+    def to_s
+      "Permissions: {:read => #{@read}, :write => #{@write}, :manage => #{@manage} , :delete => #{@delete}, :create => #{@create}, :get => #{@get}, :update => #{@update}, :join => #{@join}}"
+    end
+
+    def calculate_bitmask
+      sum = 0
+
+      if @read
+        sum |= 1
+      end
+
+      if @write
+        sum |= 2
+      end
+
+      if @manage
+        sum |= 4
+      end
+
+      if @delete
+        sum |= 8
+      end
+
+      if @create
+        sum |= 16
+      end
+
+      if @get
+        sum |= 32
+      end
+
+      if @update
+        sum |= 64
+      end
+
+      if @join
+        sum |= 128
+      end
```

Comment:
@kleewho all these can be much smaller:
```ruby
sum |= 1 if @read
sum |= 2 if @write
sum |= 4 if @manage
sum |= 8 if @delete
sum |= 16 if @create
sum |= 32 if @get
sum |= 64 if @update
sum |= 128 if @join
```

And pretty easy to read :)

Reply:
definitely better